### PR TITLE
support in query condition

### PR DESCRIPTION
**Pull request by LuciferYang**
support "column in (v1,v2..,vn)" to use index

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala [hunk 1]
```diff
@@ -112,6 +112,10 @@ private[oap] object ScannerBuilder extends Logging {
         val leftMap = optimizeFilterBound(leftFilter, ic)
         val rightMap = optimizeFilterBound(rightFilter, ic)
         combineIntervalMaps(leftMap, rightMap, ic, needMerge = false)
+      case In(attribute, keys) =>
+        val eqBounds = keys.flatMap(key => optimizeFilterBound(EqualTo(attribute, key), ic))
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: Why don't we use `key => new RangeInterval(key, key, true, ture)` and append it to ArrayBuffer?
Call `optimizeFilterBound` means ArrayBuffer -> HashMap -> ArrayBuffer. Looks weird.

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala [hunk 2]
```diff
@@ -394,4 +394,83 @@ class FilterSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEac
       Row(2, "this is test 2") :: Row(3, "this is test 3") :: Nil)
     sql("drop oindex index1 on oap_test")
   }
+
+  test("testInCondtions") {
```
Comment (linhong-intel) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala]: I think we need some tests for OAP as well.

Comment (LuciferYang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala]: ok,I'll add test case for oap

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala (hunk 1 above)
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: do like keys.map(any => {
          val key = any.asInstanceOf[Key]
          new RangeInterval(key, key, true, true)
        }).to[ArrayBuffer] ?

Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: if do like this, will throw exception: java.lang.String cannot be cast to org.apache.spark.sql.catalyst.InternalRow

Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: shall we do like this  
val eqBounds = keys.distinct
          .map(value => {
          val key = InternalRow(CatalystTypeConverters.convertToCatalyst(value))
          new RangeInterval(key, key, true, true)
        }).to[ArrayBuffer]

Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: or val key = ic.unapply(value).get?

Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: OK, I see... The `unapply` ...
All of them seem no better than your first one. I'm OK with keeping current one now.

Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: so I was very tangled, maybe we should add a new unapply method  
 def unapply(values: Array[Any]): Option[Array[Key]] =
    Some(values.map(value => InternalRow(CatalystTypeConverters.convertToCatalyst(value)))),
then we can use key => new RangeInterval(key, key, true, ture) and trans to arraybuffer with one row

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala [hunk 3]
```diff
@@ -112,6 +111,11 @@ private[oap] object ScannerBuilder extends Logging {
         val leftMap = optimizeFilterBound(leftFilter, ic)
         val rightMap = optimizeFilterBound(rightFilter, ic)
         combineIntervalMaps(leftMap, rightMap, ic, needMerge = false)
+      case In(attribute, ic(keys)) =>
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: I'm not sure `unapply` make sense if there is no corresponding `apply`.  That's why I didn't ask you to do this.
Anyway, The code looks much neater here and looks good to me.

**Comment by linhong-intel** [comment]
lgtm

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala [hunk 4]
```diff
@@ -207,6 +207,9 @@ private[oap] class IndexContext(meta: DataSourceMeta) extends Logging {
   def unapply(value: Any): Option[Key] =
     Some(InternalRow(CatalystTypeConverters.convertToCatalyst(value)))
 
+  def unapply(values: Array[Any]): Option[Array[Key]] =
+    Some(values.map(value => InternalRow(CatalystTypeConverters.convertToCatalyst(value))))
+
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: Nit: extra empty line

Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: del it....

**Comment by adrian-wang** [comment]
LGTM pending test, thanks!

**Pull request merged**